File: src/main/java/cs/ut/config/nodes/DirectoryPathConfiguration.java
package cs.ut.config.nodes;

import cs.ut.exceptions.NirdizatiRuntimeException;
import org.apache.commons.io.FileUtils;

import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;
import java.io.File;
import java.io.IOException;

@XmlRootElement(name = "paths")
public class DirectoryPathConfiguration {

    @XmlElement(name = "python")
    private String python;

    @XmlElement(name = "userLogDirectory")
    private String userLogDirectory;

    @XmlElement(name = "userModelDirectory")
    private String userModelDirectory;

    @XmlElement(name = "scriptDirectory")
    private String scriptDirectory;

    @XmlElement(name = "datasetDirectory")
    private String datasetDirectory;

    @XmlElement(name = "trainDirectory")
    private String trainDirectory;

    @XmlElement(name = "pklDirectory")
    private String pklDirectory;

    @XmlElement(name = "ohpdir")
    private String ohpdir;

    @XmlElement(name = "detailedDir")
    private String detailedDir;

    @XmlElement(name = "featureDir")
    private String featureDir;

    @XmlElement(name = "validationDir")
    private String validationDir;

    @XmlElement(name = "tmpDir")
    private String tmpDir;

    public String getUserLogDirectory() {
        return userLogDirectory;
    }

    public String getUserModelDirectory() {
        return userModelDirectory;
    }

    public String getScriptDirectory() {
        return scriptDirectory;
    }

    public String getDatasetDirectory() {
        return datasetDirectory;
    }

    public String getTrainDirectory() {
        return trainDirectory;
    }

    public String getPklDirectory() {
        return pklDirectory;
    }

    public String getOhpdir() {
        return ohpdir;
    }

    public String getPython() {
        return python;
    }

    public String getDetailedDir() {
        return detailedDir;
    }

    public String getFeatureDir() {
        return featureDir;
    }

    public String getValidationDir() {
        return validationDir;
    }

    public String getTmpDir() {
        return tmpDir;
    }

    public void validatePathsExist() {
        File file = new File(userLogDirectory);
        createDirIfAbsent(file);

        file = new File(tmpDir);
        if (file.exists()) {
            try {
                FileUtils.deleteDirectory(file);
            } catch (IOException e) {
                throw new NirdizatiRuntimeException("No permission to delete tmp dir");
            }
        }
        if (!file.mkdir()) {
            throw new NirdizatiRuntimeException("No permission to create tmp dir");
        }

        file = new File(userModelDirectory);
        createDirIfAbsent(file);

        file = new File(datasetDirectory);
        createDirIfAbsent(file);

        file = new File(trainDirectory);
        createDirIfAbsent(file);

        file = new File(scriptDirectory.concat("core/").concat(datasetDirectory));
        createDirIfAbsent(file);

        file = new File(scriptDirectory.concat("core/").concat(trainDirectory));
        createDirIfAbsent(file);

        file = new File(scriptDirectory.concat(pklDirectory));
        createDirIfAbsent(file);

        file = new File(scriptDirectory.concat("core/").concat(ohpdir));
        createDirIfAbsent(file);
    }

    protected void createDirIfAbsent(File file) {
        if (!file.exists()) {
            file.mkdir();
        }
    }
}

File: src/main/kotlin/cs/ut/config/nodes/DirectoryPathConfiguration.kt
package cs.ut.config.nodes

import cs.ut.exceptions.NirdizatiRuntimeException
import org.apache.commons.io.FileUtils
import java.io.File
import javax.xml.bind.annotation.XmlAccessType
import javax.xml.bind.annotation.XmlAccessorType
import javax.xml.bind.annotation.XmlElement
import javax.xml.bind.annotation.XmlRootElement

@XmlRootElement(name = "paths")
@XmlAccessorType(XmlAccessType.FIELD)
class DirectoryPathConfiguration(

        @XmlElement(name = "python")
        var python: String,

        @XmlElement(name = "userLogDirectory")
        var userLogDirectory: String,

        @XmlElement(name = "userModelDirectory")
        var userModelDirectory: String,

        @XmlElement(name = "scriptDirectory")
        var scriptDirectory: String,

        @XmlElement(name = "datasetDirectory")
        var datasetDirectory: String,

        @XmlElement(name = "trainDirectory")
        var trainDirectory: String,

        @XmlElement(name = "pklDirectory")
        var pklDirectory: String,

        @XmlElement(name = "ohpdir")
        var ohpdir: String,

        @XmlElement(name = "detailedDir")
        var detailedDir: String,

        @XmlElement(name = "featureDir")
        var featureDir: String,

        @XmlElement(name = "validationDir")
        var validationDir: String,

        @XmlElement(name = "tmpDir")
        var tmpDir: String) {

    constructor() : this("", "", "", "", "", "", "", "", "", "", "", "")

    fun createTmpDir() {
        val tmpDir = File(tmpDir)
        if (tmpDir.exists()) {
            FileUtils.deleteDirectory(tmpDir)
        }

        if (!tmpDir.mkdir()) {
            throw NirdizatiRuntimeException("No permission to create tmp directory")
        }
    }
}